File: 2/2.ts
/**
 * Definition for singly-linked list.
 */
 class ListNode {
    val: number
    next: ListNode | null
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.next = (next===undefined ? null : next)
    }
}

function addTwoNumbers(l1: ListNode | null, l2: ListNode | null): ListNode | null {
    const dummy = new ListNode();
    let node = dummy;

    let carry = 0;
    while (l1 || l2) {
        const sum = carry + (l1 ? l1.val : 0) + (l2 ? l2.val : 0);
        const digit = sum % 10;
        carry = Math.floor(sum / 10);

        node.next = new ListNode(digit);
        node = node.next;

        l1 = l1?.next;
        l2 = l2?.next;
    }
    if (carry !== 0) {
        node.next = new ListNode(carry);
    }

    return dummy.next;
};

const solution1 = addTwoNumbers;

export {};

File: 2/2.hs
data ListNode = Nil
              | ListNode
              { val  :: Int
              , next :: ListNode
              }

newListNode :: [Int] -> ListNode
newListNode []     = Nil
newListNode (x:xs) = ListNode x (newListNode xs)

toList :: ListNode -> [Int]
toList Nil               = []
toList (ListNode v next) = v : toList next

addTwoNumbers :: ListNode -> ListNode -> ListNode
addTwoNumbers l1 l2 = add l1 l2 0
  where
    add :: ListNode -> ListNode -> Int -> ListNode
    add Nil Nil 0 = Nil
    add Nil Nil c = ListNode c Nil
    add a   b   c =
      let
        x = case a of
          Nil -> 0
          ListNode vx _ -> vx

        y = case b of
          Nil -> 0
          ListNode vy _ -> vy

        total = x + y + c
        digit = total `mod` 10
        carry = total `div` 10

        nextA = case a of
          Nil -> Nil
          ListNode _ na -> na

        nextB = case b of
          Nil -> Nil
          ListNode _ nb -> nb

      in ListNode digit (add nextA nextB carry)
